### PR TITLE
Fix for setting ssl info in stream_info when starttls transport socket is used

### PR DESCRIPTION
**Pull request by VishalDamgude**
Commit Message: Fix for setting ssl info in stream_info when starttls transport socket is used
Additional Description:
When starttls transport socket is configured either for listener or for upstream cluster, it is initialized with plain text (raw buffer) socket. i.e, ssl info is not populated (=null) when socket is initialized. So, the stream_info object also has empty ssl info, which is a problem if we want to print it in access logs.
The SSL handshake takes place only after socket is switched to TLS by protocol specific command(STARTTLS). So we need to explicitly set ssl info in stream info for both downstream and upstream after socket is switched to TLS. This PR addresses this problem.
Any suggestions on the approach are welcome!

Risk Level: Low
Testing: Tested with smtp filter(locally) with starttls command.
Docs Changes: No
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue] https://github.com/envoyproxy/envoy/issues/26344
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/common/tcp_proxy/tcp_proxy.cc [hunk 1]
```diff
@@ -682,7 +682,15 @@ Network::FilterStatus Filter::onNewConnection() {
   return establishUpstreamConnection();
 }
 
-bool Filter::startUpstreamSecureTransport() { return upstream_->startUpstreamSecureTransport(); }
+bool Filter::startUpstreamSecureTransport() {
+  bool switched_to_tls = upstream_->startUpstreamSecureTransport();
+  if (switched_to_tls) {
+    StreamInfo::UpstreamInfo& upstream_info = *getStreamInfo().upstreamInfo();
+    auto ssl_info = upstream_->getUpstreamConnectionSslInfo();
```
Comment (lizan) [inline source/common/tcp_proxy/tcp_proxy.cc:689]: not used?

File: source/common/tcp_proxy/upstream.cc [hunk 2]
```diff
@@ -49,6 +49,13 @@ bool TcpUpstream::startUpstreamSecureTransport() {
              : upstream_conn_data_->connection().startSecureTransport();
 }
 
+Ssl::ConnectionInfoConstSharedPtr TcpUpstream::getUpstreamConnectionSslInfo() {
+  if (!upstream_conn_data_) {
```
Comment (lizan) [inline source/common/tcp_proxy/upstream.cc:53]: prefer `!= nullptr`

**Comment by mattklein123** [comment]
Friendly ping @lizan

**Comment by adisuissa** [comment]
@lizan can you PTAL?

**Comment by VishalDamgude** [comment]
@cpakulski Can you also please take a look at this?

**Comment by cpakulski** [comment]
Sure. Looking into it.

File: envoy/tcp/upstream.h [hunk 3]
```diff
@@ -144,6 +144,11 @@ class GenericUpstream {
    * to secure mode. Implemented only by start_tls transport socket.
    */
   virtual bool startUpstreamSecureTransport() PURE;
+
+  /* Called when upstream starttls socket is converted to tls and upstream ssl info
+   *  needs to be set in connection's stream_info.
+   */
+  virtual Ssl::ConnectionInfoConstSharedPtr getUpstreamConnectionSslInfo() PURE;
```
Comment (cpakulski) [inline envoy/tcp/upstream.h:151]: I think you can do it a bit simpler and avoid adding another accessor (`getUpstreamConnectionSslInfo`). Just modify `GenericUpstream::startUpstreamSecureTransport` in https://github.com/envoyproxy/envoy/blob/main/envoy/tcp/upstream.h#L146 to return Tcp::ConnectionPool::ConnectionData* (as in https://github.com/envoyproxy/envoy/blob/main/envoy/tcp/upstream.h#L140-L141). Then in https://github.com/envoyproxy/envoy/blob/main/source/common/tcp_proxy/upstream.cc#L46-L50 return valid pointer on success or nullptr on failure.

In `Filter::startUpstreamSecureTransport` you can check the return value and extract ssl() directly.

WDYT?

File: source/common/tcp_proxy/tcp_proxy.cc [hunk 4]
```diff
@@ -682,7 +682,14 @@ Network::FilterStatus Filter::onNewConnection() {
   return establishUpstreamConnection();
 }
 
-bool Filter::startUpstreamSecureTransport() { return upstream_->startUpstreamSecureTransport(); }
+bool Filter::startUpstreamSecureTransport() {
+  bool switched_to_tls = upstream_->startUpstreamSecureTransport();
```
Comment (cpakulski) [inline source/common/tcp_proxy/tcp_proxy.cc:686]: If you modify signature of this method as described above, you can get access to Pool's connection, extract ssl and  set it in upstream_info.

**Review by cpakulski — CHANGES_REQUESTED** [review]
Generally OK, but I think we can simplify it as described in code review comments.

**Review by cpakulski — DISMISSED** [review]
Code looks OK to me.
We just need to make sure that CI completes successfully,

**Comment by cpakulski** [comment]
@phlax , would you mind approving CI run? Thanks.

**Comment by yanavlasov** [comment]
I think the CI is stuck and I'm unable to restart it. Please merge main to kick the CI again.

**Comment by phlax** [comment]
i tried kicking the CI manually and that doesnt seem to have worked, i guess it does need a main merge

**Comment by VishalDamgude** [comment]
Hi @phlax ,
I have merged with main branch. Please start CI run.

File: envoy/tcp/upstream.h [hunk 5]
```diff
@@ -144,6 +144,11 @@ class GenericUpstream {
    * to secure mode. Implemented only by start_tls transport socket.
    */
   virtual bool startUpstreamSecureTransport() PURE;
+
+  /* Called when upstream starttls socket is converted to tls and upstream ssl info
```
Comment (lizan) [inline envoy/tcp/upstream.h:148]: nit: comment format.

**Review by lizan — COMMENTED** [review]
just one nit. @phlax feel free to merge when the comment format is resolved.

File: envoy/tcp/upstream.h (hunk 5 above)
Comment (VishalDamgude) [reply envoy/tcp/upstream.h]: Updated comment format

**Review by phlax — APPROVED** [review]
lgtm, thanks @VishalDamgude

**Pull request merged**